dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java

```diff
@@ -97,20 +109,55 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
 
     @Override
     protected void afterDeploymentValidation(@Observes AfterDeploymentValidation done, BeanManager beanManager) {
-        // Catch DeploymentException that is thrown if there are any validation errors with the Reactive Messaging Configuration
-        try {
-            MediatorManager mediatorManager = configureMediatorManager(beanManager);
+        final ContextBeginnerEnder contextBeginnerEnder = createContextBeginnerEnderIfCheckpointEnabled();
+        final MediatorManager mediatorManager = configureMediatorManager(beanManager); //TODO I only fonud one DeploymentException in the small rye source so there's no need for a second catch (DeploymentException) block. But double check this with the team. 
+
+        CheckpointHook checkpointHook = new CheckpointHook() {
+
+            @Override
+            public void restore() {
+                try (ContextBeginnerEnder cbe = contextBeginnerEnder.beginContext()) {
+                    startMediatorManager(mediatorManager);
+                }
+            }
+        };
+
+        // If checkpoint is enabled this if statement will register the hook we created above, so startMediatorManager() will be called after restore. Then it returns true.
+        // If checkpoint is disabled this if statement will return false, so startMediatorManager() will be called right away.
+        if (!CheckpointPhase.getPhase().addMultiThreadedHook(checkpointHook)) {
```

Comment:
Given there is a Phase check that is run as part of `createContextBeginnerEnderIfCheckpointEnabled`. would `CheckpointPhase.getPhase() == CheckpointPhase.INACTIVE` achieve the same as this check, as that is more understandable, or are there some other phases that we need to be account for that this check also covers

Reply:
There actually is a reason for this, but its complicated. The alternative would be 

```if (phase != CheckpointPhase.INACTIVE && !phase.restored() && phase.addMultiThreadedHook(myHook)){
    // this is the checkpoint side
} else {
    // this is the normal case (or after restore)
}```

However Tom Watson warns me that `phase.restored()` has threading risks and recommended the code used here.

Reply:
If that is what Tom is recommending then that is what we should follow.

Reply:
You could avoid creating the hook with some checks for `!=CheckpointPhase.INACTIVE` but I don't believe that would improve performance and I think it would complicate the code a bit more than necessary.  I think what is here is good.